In [1]:
import pandas as pd

CARGA Y CONSOLIDACION DEL DATASET

In [2]:

#cargamos el dataset
df_2009 = pd.read_csv("data/2009_1perc.csv", encoding= "latin-1")
df_2010 = pd.read_csv("data/2010_1perc.csv", encoding= "latin-1")

#consolidamos ambos set
# Verificar que ambos DataFrames tienen la misma estructura de columnas
if list(df_2009.columns) == list(df_2010.columns):
    # Concatenar los DataFrames
    df_consolidado = pd.concat([df_2009, df_2010], ignore_index=True)
    print('Los archivos han sido consolidados.')
else:
    print('Los DataFrames tienen estructuras de columnas diferentes')

Los archivos han sido consolidados.


In [3]:
df_consolidado.to_csv("data/consolidado.csv", index=False)

In [4]:
df_consolidado.info()

<class 'pandas.DataFrame'>
RangeIndex: 11825 entries, 0 to 11824
Columns: 112 entries, Unnamed: 0 to detailcm
dtypes: float64(1), int64(16), object(3), str(92)
memory usage: 11.9+ MB


EXPLORACION BASICA

Obtenemos información básica del dataset: número de filas/columnas, tipos de datos, estadísticas básicas y número de vacios por variable.

In [ ]:
print(f"El DataFrame consolidado tiene {df_consolidado.shape[0]} filas y {df_consolidado.shape[1]} columnas.")

object_count = (df_consolidado.dtypes == 'object').sum()
print(f"Numero de variables tipo object: {object_count}")
object_count = (df_consolidado.dtypes == 'str').sum()
print(f"Numero de variables tipo string: {object_count}")
# Contar cuántas variables son de tipo 'integer'
integer_count = (df_consolidado.dtypes == 'int').sum() + (df_consolidado.dtypes == 'int64').sum()
print(f"Numbero de variables tipo entero: {integer_count}")
# Contar cuántas variables son de tipo 'float'
float_count = (df_consolidado.dtypes == 'float').sum() + (df_consolidado.dtypes == 'float64').sum()
print(f"Numbero de variables tipo float: {float_count}")
missing_values = df_consolidado.isnull().sum()
print(f"Numbero valores perdidos: {missing_values}")
print("-------------")
print("Estadisticos del dataset")
df_consolidado.describe().T

El DataFrame consolidado tiene 11825 filas y 112 columnas.
Numero de variables tipo object: 3
Numero de variables tipo string: 92
Numbero de variables tipo entero: 32
Numbero de variables tipo float: 2
Numbero valores perdidos: Unnamed: 0    0
year          0
pct           0
ser_num       0
datestop      0
             ..
xcoord        0
ycoord        0
dettypcm      0
linecm        0
detailcm      0
Length: 112, dtype: int64
-------------
Estadisticos del dataset


,Unnamed: 0,year,pct,ser_num,datestop,timestop,perobs,perstop,compyear,comppct,repcmd,revcmd,dob,age,ht_feet,ht_inch,weight
count,11825.000000,11825.000000,11825.000000,11825.000000,1.182500e+04,11825.000000,11825.000000,11825.000000,11825.0,11825.0,11825.000000,11825.000000,1.182500e+04,11825.000000,11825.000000,11825.000000,11825.000000
mean,295983.814799,2009.508499,68.625624,5697.994334,6.364221e+06,1413.725497,2.654391,5.649556,0.0,0.0,243.354926,238.986385,7.316195e+06,28.968541,5.183340,6.408034,170.093615
std,170820.705702,0.499949,33.037826,5152.486022,3.429091e+06,744.726359,9.759031,9.758535,0.0,0.0,297.788599,299.394740,3.729206e+06,29.205278,0.398801,3.405503,37.195884
min,52.000000,2009.000000,1.000000,1.000000,1.012009e+06,0.000000,0.000000,1.000000,0.0,0.0,1.000000,1.000000,1.011900e+06,1.000000,3.000000,0.000000,1.000000
25%,147515.000000,2009.000000,42.000000,2002.000000,3.232009e+06,1000.000000,1.000000,3.000000,0.0,0.0,70.000000,66.000000,4.121982e+06,19.000000,5.000000,4.000000,150.000000
50%,296652.000000,2010.000000,73.000000,4366.000000,6.162010e+06,1605.000000,1.000000,5.000000,0.0,0.0,105.000000,103.000000,7.261982e+06,25.000000,5.000000,7.000000,170.000000
75%,443097.000000,2010.000000,102.000000,7840.000000,9.232009e+06,2035.000000,3.000000,5.000000,0.0,0.0,177.000000,176.000000,1.105199e+07,34.000000,5.000000,9.000000,180.000000
max,601281.000000,2010.000000,123.000000,31694.000000,1.231201e+07,2830.000000,926.000000,857.000000,0.0,0.0,879.000000,879.000000,1.231200e+07,999.000000,7.000000,11.000000,999.000000


Algunas observaciones acerca del dataset:

La varible "age" está en unidades de años, por lo que algunos datos de la variable "age" son incosistentes como el max= 999
La varible "weight" está en kilos o libras por lo que algunos datos de la variable "weight" son inconsistentes como max= 999

Se eliminan las siguientes variables ya que son de identificación única (no aportan generalización)
ser_num	ID único
officrid	ID del oficial
addrnum	    Número exacto dirección
premname	Nombre específico lugar
aptnum	    Apartamento
xcoord	    Coordenada exacta
ycoord	    Coordenada exacta

Se eliminan  las siguientes variables ya que tienen alta granularidad textual
stname	    Miles de calles
crossst	    Alta cardinalidad
stinter	    Alta cardinalidad
othfeatr	Texto libre
crimsusp	Texto ambiguo
ac_other	Texto ambiguo
cs_other	Texto libre
sb_other	Texto libre

Se eliminan variables redundantes o poco utiles
dob	 Ya existe age
zip	 Muy granular
post Poco valor predictivo




In [7]:
cols_to_drop = [
    "ser_num", "officrid", "arstoffn", "sumissue", "sumoffen",
    "revcmd", "detailcm", "linecm", "dettypcm",
    "addrnum", "premname", "aptnum", "xcoord", "ycoord",
    "dob", "stname", "crossst", "stinter", "othfeatr",
    "cs_other", "sb_other", "ac_other"
]
df_consolidado.drop(columns=cols_to_drop, inplace=True)

df_consolidado.info()

<class 'pandas.DataFrame'>
RangeIndex: 11825 entries, 0 to 11824
Data columns (total 90 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  11825 non-null  int64  
 1   year        11825 non-null  int64  
 2   pct         11825 non-null  int64  
 3   datestop    11825 non-null  int64  
 4   timestop    11825 non-null  int64  
 5   recstat     11825 non-null  str    
 6   inout       11825 non-null  str    
 7   trhsloc     11825 non-null  str    
 8   perobs      11825 non-null  float64
 9   crimsusp    11825 non-null  str    
 10  perstop     11825 non-null  int64  
 11  typeofid    11825 non-null  str    
 12  explnstp    11825 non-null  str    
 13  othpers     11825 non-null  str    
 14  arstmade    11825 non-null  str    
 15  compyear    11825 non-null  int64  
 16  comppct     11825 non-null  int64  
 17  offunif     11825 non-null  str    
 18  frisked     11825 non-null  str    
 19  searched    11825 non-null  str    
